flutter_install_type.go

```diff
@@ -0,0 +1,279 @@
+	after3_0_0 = false
+	after3_2_1 = false
+	regex := regexp.MustCompile(`\d+\.\d+\.\d+`)
+	versionParts := strings.Split(regex.FindString(versionOut), ".")
+	if len(versionParts) >= 3 {
+		var major, minor, patch int
+		_, majorErr := fmt.Sscan(versionParts[0], &major)
+		_, minorErr := fmt.Sscan(versionParts[1], &minor)
+		_, patchErr := fmt.Sscan(versionParts[2], &patch)
+
+		if majorErr == nil && minorErr == nil && patchErr == nil {
+			if major < 3 {
+				return
+			}
+
+			after3_0_0 = true                                                // FVM 3.0.0 and above
+			after3_1_0 = major > 3 || minor > 0 || (minor == 1 && patch > 0) // FVM 3.1.0 and above
+			after3_2_1 = major > 3 || minor > 2 || (minor == 2 && patch > 0) // FVM 3.2.1 and above
+
+			return
+		}
+		err = fmt.Errorf("parse fvm version: %s:\nmajor:%w\nminor: %w\npatch: %w", versionOut, majorErr, minorErr, patchErr)
+	} else {
+		err = fmt.Errorf("parse fvm version: %s", versionOut)
+	}
+
+	return
+}
+
+func fvmCreateVersionString(version flutterVersion) string {
+	versionString := version.version
+	if versionString != "" {
+		if version.channel != "" {
+			versionString += "@" + version.channel
+		}
+	} else if version.channel != "" {
+		versionString = version.channel
+	} else {
+		versionString = "stable" // default to stable if no version or channel is specified
+	}
+	return versionString
+}
+
+func (f *FlutterInstaller) NewFlutterInstallTypeASDF() FlutterInstallType {
+	cmd := f.CmdFactory.Create("asdf", []string{"plugin-list"}, nil)
+	f.Donef("$ %s", cmd.PrintableCommandArgs())
+	out, err := cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil || !strings.Contains(out, "flutter") {
+		f.Warnf("asdf version manager is not available")
+		return FlutterInstallType{
+			Name:        ASDFName,
+			IsAvailable: false,
+		}
+	}
+
+	return FlutterInstallType{
+		Name:        ASDFName,
+		IsAvailable: true,
+		InstalledVersionsCommand: func() *command.Command {
+			cmd := f.CmdFactory.Create("asdf", []string{"list", "flutter"}, nil)
+			f.Donef("$ %s", cmd.PrintableCommandArgs())
+			return &cmd
+		},
+		Install: func(version flutterVersion) error {
+			cmd := f.CmdFactory.Create("asdf", []string{"install", "flutter", asdfCreateVersionString(version)}, nil)
+			f.Donef("$ %s", cmd.PrintableCommandArgs())
+			if out, err := cmd.RunAndReturnTrimmedCombinedOutput(); err != nil {
+				return fmt.Errorf("install: %s %s", err, out)
+			} else {
+				f.Debugf("Installed Flutter: %s", out)
+				shimsPath := os.Getenv("HOME") + ASDFShimsPath
+				path := os.Getenv("PATH")
+				if err := os.Setenv("PATH", fmt.Sprintf("%s:%s", shimsPath, path)); err != nil {
+					return fmt.Errorf("set env: %s", err)
+				}
+				f.Debugf("Added asdf shims to PATH: %s", os.Getenv("PATH"))
+			}
+			return nil
+		},
+		SetDefault: func(version flutterVersion) error {
+			versionString := asdfCreateVersionString(version)
+			cmd := f.CmdFactory.Create("asdf", []string{"reshim", "flutter", versionString}, nil)
```

Comment:
Logically, this belongs at the end of `Install()` I think. And I'm not even sure this is needed in the case of Flutter, but you might know better